### DISCUSSION TITLE
Is there an API for what MemberNotNullWhenAttribute does?

### DISCUSSION THREAD

**Discussion by bfriesen** [dotnet/roslyn#68035] [Q&A]
I'm looking for a public API that does the sort of analysis that `MemberNotNullWhenAttribute` triggers, where it looks at the local code to see if a member has been verified to have a true or false value. In the case of `MemberNotNullWhenAttribute`, the compiler uses that information to know that a value is no longer 'maybe null' and is now 'not null'. In my case, I want to use that information to report a custom diagnostic from my analyzer.

The following struct demonstrates the kind of analysis I'm looking for. What I really want to do is report a custom diagnostic when `Value` is accessed without first verifying that `HasValue` is true. What this does instead is report a 'maybe null' warning when `Value` is accessed without first verifying that `HasValue` is true.

```c#
struct Result<T>
{
    private readonly bool _hasValue;
    private readonly T _value;

    public Result(T value)
    {
        _value = value;
        _hasValue = value != null;
    }
    
    [MemberNotNullWhen(true, nameof(Value))]
    public bool HasValue => _hasValue;

    public T? Value =>
        _hasValue
            ? _value
            : throw new InvalidOperationException();
}
```

In this example, when `T` is a reference type, the `Value` property is 'maybe null' by default because it is has a type of `T?`. So the following method gives a compiler warning because we're trying to assign a 'maybe null' string to a 'non null' string.

![image](https://user-images.githubusercontent.com/1056340/235471782-3e9235a8-8e92-46e7-b76c-9b05a0ae78f0.png)

Of course, if `T` is <em>not</em> a reference type, we will never get a compiler warning from the `Value` property. This is why using `MemberNotNullWhenAttribute` directly won't work for my use-case.

![image](https://user-images.githubusercontent.com/1056340/235475391-b41b41dc-1966-41e5-b1b5-85576179f968.png)

We can make the warning go away by making sure the `HasValue` property is true before accessing the `Value` property.

![image](https://user-images.githubusercontent.com/1056340/235473163-87417e7f-17da-4432-a748-a2baa4fad5bf.png)

There are, of course, many ways of ensuring that `HasValue` is true, and the compiler has no trouble finding them. It is this ability to find any possible way of verifying that `HasValue` is true that I'm after.

![image](https://user-images.githubusercontent.com/1056340/235473416-98ceb830-7d63-4b93-a2a9-b71aba61c3b9.png)

Is there a public API that does this sort of analysis? Or if not, can someone point me to where this analysis is actually performed by the compiler?

**Comment by jasonmalinowski** [comment]
The nullable flow state is exposed if you call GetTypeInfo() on certain nodes, and you can figure out how some things are annotated with other calls too. It'd be helpful to understand what you're trying to do here so we can give better help though. Although I'm a bit confused, are you trying to understand the nullable information really, or just how the compiler determines what a "it's true" check looks like?

> Or if not, can someone point me to where this analysis is actually performed by the compiler?

The analysis is very complex. :smile: https://github.com/dotnet/roslyn/blob/main/src/Compilers/CSharp/Portable/FlowAnalysis/NullableWalker.cs if you're curious, but you should really try to explain what you want to do here so we can point you to the right APIs since reimplementing that is a non-starter.

**Reply by bfriesen** [reply]
I'm not interested in the actual nullable flow state, but instead a subset of the analysis that it does. What I have now is a `Result<T>` struct, similar to this:

```c#
struct Result<T>
{
    private readonly bool _hasValue;
    private readonly T _value;

    public Result(T value)
    {
        _value = value;
        _hasValue = value != null;
    }
    
    public bool HasValue => _hasValue;

    public T Value =>
        _hasValue
            ? _value
            : throw new InvalidOperationException();
}
```

As you can see, a program cannot successfully access the `Value` property of a `Result<T>` when the `HasValue` property is false. I would like to write an analyzer that enforces this. For example, the following block should report a diagnostic around `[|result.Value|]` because we don't know that `HasValue` is true.

```c#
void Example1(Result<string> result)
{
    Console.WriteLine([|result.Value|]);
}
```

This example should not report a diagnostic because it verifies that `HasValue` is true before accessing the `Value` property:

void Example2(Result<string> result)
{
    if (result.HasValue)
        Console.WriteLine(result.Value);
}

---

In my analyzer, it's trivial to find usages of the `Result<T>.Value` property. What I need to do next - the hard part - is determine if the same result's `HasValue` has been verified to be true at that point in the method. While brainstorming, it occurred to me that usage of the `MemberNotNullWhenAttribute` initiates a very similar analysis (as the `Result<T>` struct in my original question demonstrates). In both cases, a default assumption is made. In my original type, the assumption is that `Value` is assumed to be 'maybe null' and the compiler will mark violations (such as assigning to a 'not null' variable) as such, *unless* `HasValue` has been determined to be true first. In this post's type, the assumption is that the `Value` property is unsafe to access, *unless* `HasValue` has been determined to be true first.

It would be *wonderful* if there was an API for answering the question "has it been determined that a member of this variable/parameter/etc has a value of true/false?"

**Reply by bfriesen** [reply]
That said, pointing me to a walker of some sort is about what I expected. I thought maybe there would be a public API somewhere that maybe inherits `OperationWalker` that collects the answer as it visits operation.
